Simulator/Layouts/Layout.cpp

```diff
@@ -94,10 +78,10 @@ void Layout::registerGraphProperties()
 void Layout::setup()
 {
    // Allocate memory
-   xloc_ = new VectorMatrix(MATRIX_TYPE, MATRIX_INIT, 1, numVertices_);
-   yloc_ = new VectorMatrix(MATRIX_TYPE, MATRIX_INIT, 1, numVertices_);
-   dist2_ = new CompleteMatrix(MATRIX_TYPE, MATRIX_INIT, numVertices_, numVertices_);
-   dist_ = new CompleteMatrix(MATRIX_TYPE, MATRIX_INIT, numVertices_, numVertices_);
+   xloc_ = make_unique<VectorMatrix>(MATRIX_TYPE, MATRIX_INIT, 1, numVertices_);
+   yloc_ = make_unique<VectorMatrix>(MATRIX_TYPE, MATRIX_INIT, 1, numVertices_);
+   dist2_ = make_unique<CompleteMatrix>(MATRIX_TYPE, MATRIX_INIT, numVertices_, numVertices_);
+   dist_ = make_unique<CompleteMatrix>(MATRIX_TYPE, MATRIX_INIT, numVertices_, numVertices_);
```

Comment:
We should probably talk about whether the data members should be pointers to these objects or just these objects themselves. I don't remember why they were pointers in the first place...

Reply:
One thing that comes to mind is that we don't allocate memory for these data members when the Layout is instantiated because at that point we don't yet know the number of vertices for the simulation. That might be the reason these are pointers and not objects.
By the way, the delayed construction of class member objects is another potential use for `std::optional`.

Reply:
That's a possibility, and at some level the question comes down to what approach best communicates the purpose of the code ("what going on?") to a human reader, since the various approaches all "work". I don't have the answer right now, but I'll propose a third possibility: declare the data members with default sizes (like 1x1 — or modify the classes to add defaults to the constructors), and then, when we know how big they should be, just use assignment to change them. This would be like using an `int` data member, for example — it might be initialized to zero or whatever in the constructor, but in some other method set to a "better" value.